[export] Exempt autograd ops for predispatch export

andrewlee302 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116527
* #116339

Summary:
We intend to preserve autograd ops for predispatch export. Therefore, we
need to exempt the autograd ops in some places, e.g. verifier and
proxy_tensor.py.

Test Plan:
python test/export/test_export.py -k test_predispatch_export_with_autograd_op

andrewlee302 committed:
[export] Exempt autograd ops for predispatch export

Summary:
We intend to preserve autograd ops for predispatch export. Therefore, we
need to exempt the autograd ops in some places, e.g. verifier and
proxy_tensor.py.

Test Plan:
python test/export/test_export.py -k test_predispatch_export_with_autograd_op

[ghstack-poisoned]

tugsbayasgalan commented on torch/_export/verifier.py:
```diff
@@ -166,6 +166,8 @@ def _allowed_op_types() -> Tuple[Type[Any], ...]:
                 torch.sym_not,
                 torch.sym_sqrt,
+                # Predispatch export is able to contain autograd ops.
+                torch._C._set_grad_enabled
```
maybe not needed?

tugsbayasgalan commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -569,7 +569,13 @@ def __init__(self, tracer):
         kwargs = kwargs or {}
         if func in _side_effectful_need_to_be_preserved_pre_dispatch:
-            return self.tracer.create_node("call_function", func, args, {})
+            # It's for passing the verifier which needs to verify the meta['val']
```
nit: Can you add that this is export verifier we are talking about?

andrewlee302 committed:
Update on "[export] Exempt autograd ops for predispatch export"


Summary:
We intend to preserve autograd ops for predispatch export. Therefore, we
need to exempt the autograd ops in some places, e.g. verifier and
proxy_tensor.py.

Test Plan:
python test/export/test_export.py -k test_predispatch_export_with_autograd_op

[ghstack-poisoned]

andrewlee302 commented:
@pytorchbot merge

andrewlee302 committed:
Update on "[export] Exempt autograd ops for predispatch export"


Summary:
We intend to preserve autograd ops for predispatch export. Therefore, we
need to exempt the autograd ops in some places, e.g. verifier and
proxy_tensor.py.

Test Plan:
python test/export/test_export.py -k test_predispatch_export_with_autograd_op

[ghstack-poisoned]

andrewlee302 committed:
Update on "[export] Exempt autograd ops for predispatch export"


Summary:
We intend to preserve autograd ops for predispatch export. Therefore, we
need to exempt the autograd ops in some places, e.g. verifier and
proxy_tensor.py.

Test Plan:
python test/export/test_export.py -k test_predispatch_export_with_autograd_op

[ghstack-poisoned]

andrewlee302 committed:
Update on "[export] Exempt autograd ops for predispatch export"


Summary:
We intend to preserve autograd ops for predispatch export. Therefore, we
need to exempt the autograd ops in some places, e.g. verifier and
proxy_tensor.py.

Test Plan:
python test/export/test_export.py -k test_predispatch_export_with_autograd_op

[ghstack-poisoned]

andrewlee302 commented:
@pytorchbot merge

tugsbayasgalan commented:
@pytorchbot revert -m "Need to revert this to revert the bottom diff" -c ghfirst
